gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1276 @@
+        if (response != null) {
+          return response;
+        }
+        counter++;
+      }
+    }
+    return additionsMeetDeletions(change.getAdditions(), change.getDeletions(), zone);
+    // null if everything is ok
+  }
+
+  /**
+   * Checks a rrset within a change.
+   *
+   * @param type [additions|deletions]
+   * @param index the index or addition or deletion in the list
+   * @param zone the zone that this change is applied to
+   */
+  @VisibleForTesting
+  static Response checkRrset(ResourceRecordSet rrset, ZoneContainer zone, int index, String type) {
+    if (rrset.getName() == null || !rrset.getName().endsWith(zone.zone().getDnsName())) {
+      return Error.INVALID.response(String.format(
+          "Invalid value for 'entity.change.%s[%s].name': '%s'", type, index, rrset.getName()));
+    }
+    if (rrset.getType() == null || !TYPES.contains(rrset.getType())) {
+      return Error.INVALID.response(String.format(
+          "Invalid value for 'entity.change.%s[%s].type': '%s'", type, index, rrset.getType()));
+    }
+    if (rrset.getTtl() != null && rrset.getTtl() != 0 && rrset.getTtl() < 0) {
+      return Error.INVALID.response(String.format(
+          "Invalid value for 'entity.change.%s[%s].ttl': '%s'", type, index, rrset.getTtl()));
+    }
+    if (rrset.getRrdatas() == null || rrset.isEmpty()) {
+      return Error.INVALID.response(String.format(
+          "Invalid value for 'entity.change.%s[%s].rrdata': '%s'", type, index, "<empty>"));
+    }
+    int counter = 0;
+    for (String record : rrset.getRrdatas()) {
+      if (!checkRrData(record, rrset.getType())) {
+        return Error.INVALID.response(String.format(
+            "Invalid value for 'entity.change.%s[%s].rrdata[%s]': '%s'",
+            type, index, counter, record));
+      }
+      counter++;
+    }
+    if ("deletions".equals(type)) {
+      // check that deletion has a match by name and type
+      boolean found = false;
+      for (ResourceRecordSet wrappedRrset : zone.dnsRecords().get().values()) {
+        if (rrset.getName().equals(wrappedRrset.getName())
+            && rrset.getType().equals(wrappedRrset.getType())) {
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        return Error.NOT_FOUND.response(String.format(
+            "The 'entity.change.deletions[%s]' resource named '%s (%s)' does not exist.",
+            index, rrset.getName(), rrset.getType()));
+      }
+      // if found, we still need an exact match
+      if ("deletions".equals(type)
+          && !zone.dnsRecords().get().containsValue(rrset)) {
+        // such a record does not exist
+        return Error.CONDITION_NOT_MET.response(String.format(
+            "Precondition not met for 'entity.change.deletions[%s]", index));
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Checks that for each record that already exists, we have a matching deletion. Furthermore,
```

Comment:
I was slightly confused by the comment initially.  Maybe the first sentence could be something along the lines of "checks against duplicate additions"